File: apps/desktop/src/chat/permissions.ts
interface PermissionState {
  full_access: boolean | null;
  approval_policy: string | null;
  reviewer: string | null;
}

export function permissionPreset(settings: PermissionState) {
  if (settings.full_access === true && settings.approval_policy === 'never') return 'full_access';
  if (settings.full_access === false && settings.approval_policy === 'on-request') return 'workspace';
  return null;
}

export function permissionLabel(settings: PermissionState) {
  if (settings.full_access === null || settings.approval_policy === null) return 'Permissions unavailable';
  const preset = permissionPreset(settings);
  if (preset === 'full_access') return 'Full Access';
  if (preset === 'workspace') {
    if (settings.reviewer === 'user') return 'Ask for approval';
    if (settings.reviewer === 'auto_review' || settings.reviewer === 'guardian_subagent') return 'Auto-review';
  }
  return 'Custom permissions';
}

File: crates/codex-adapter/tests/permissions.rs
use remote_codex_adapter::{desktop, events};
use remote_codex_core::session::{ApprovalsReviewer, PermissionPreset, SessionSettings};
use serde_json::json;

#[test]
fn presets_set_both_execution_and_approval_policy() {
    for (permissions, reviewer, expected, approval, reviewed_by) in [
        (
            PermissionPreset::FullAccess,
            None,
            ":danger-full-access",
            "never",
            None,
        ),
        (
            PermissionPreset::Workspace,
            Some(ApprovalsReviewer::User),
            ":workspace",
            "on-request",
            Some("user"),
        ),
        (
            PermissionPreset::Workspace,
            Some(ApprovalsReviewer::AutoReview),
            ":workspace",
            "on-request",
            Some("auto_review"),
        ),
    ] {
        let params = events::settings(
            "thread",
            SessionSettings {
                permissions: Some(permissions),
                reviewer,
                ..Default::default()
            },
        );
        assert_eq!(params["permissions"], expected);
        assert_eq!(params["approvalPolicy"], approval);
        assert_eq!(params["approvalsReviewer"].as_str(), reviewed_by);
    }
    let params = events::settings(
        "thread",
        SessionSettings {
            model: Some("selected".into()),
            ..Default::default()
        },
    );
    assert!(params.get("approvalPolicy").is_none());
    assert!(params.get("permissions").is_none());
}

#[test]
fn confirmed_settings_keep_sandbox_and_approval_independent() {
    for (approval, expected) in [
        (json!("never"), "never"),
        (json!("on-request"), "on-request"),
        (json!({"reject":{"mcp_elicitations":true}}), "custom"),
        (json!(null), "custom"),
    ] {
        let confirmed = desktop::settings(
            &json!({"sandboxPolicy":{"type":"dangerFullAccess"}, "approvalPolicy":approval}),
        );
        assert!(confirmed.full_access);
        assert_eq!(confirmed.approval_policy, expected);
    }
}
